File: ruby/gmsec4/dox/HeartbeatGenerator.rb
# Copyright 2007-2021 United States Government as represented by the
# Administrator of The National Aeronautics and Space Administration.
# No copyright is claimed in the United States under Title 17, U.S. Code.
# All Rights Reserved.

# This file is used for generating Ruby documentation; it only contains
# stubs of constructors and methods associated with the class.


module Libgmsec_ruby #:nodoc: don't document this

	# This class can be used to start a thread that will continuously publish Heartbeat messages.
	#
	# Note: Ruby can sometimes be finicky when invoking calls through the C-Ruby interface that start
	# a thread. <b>Use this class at your own risk!</b>
	#
	# ==== Example usage
	#
	#    require 'libgmsec_ruby'
	#
	#    begin
	#        config = Libgmsec_ruby::Config.new( ARGV )
	#        fields = Libgmsec_ruby::FieldArray.new
	#        fields << Libgmsec_ruby::StringField.new("MISSION-ID", "MY-MISSION")
	#        ...
	#
	#        hbgen = Libgmsec_ruby::HeartbeatGenerator.new( config, "GMSEC.MY-MISSION.MY-SAT-ID.MSG.HB.MY-COMP", 30, fields )
	#
	#        hbgen.start
	#
	#        # ... continue with other stuff as the Heartbeat Generator runs in the background as a separate thread
	#    rescue GmsecException => e
	#        puts("Error: #{e.message}")
	#    end
	#
	#
	class HeartbeatGenerator
		##
		# Constructs a Heartbeat Generator object.
		#
		# ==== Attributes
		#
		# * +config+         - The Config object to configure the ConnectionManager that is used to publish heartbeat messages.
		# * +hb_msg_subject+ - The subject to assign to the heartbeat messages.
		# * +hb_pub_rate+    - The publish rate/interval (in seconds) at which to publish heartbeat messages.
		# * +fields+         - Optional list of fields to include within the Heartbeat message.
		#
		# ==== Raises
		#
		# A GmsecException is thrown if the subject is nil or is an empty-string.
		# A GmsecException is thrown if the ConnectionManager cannot be created or cannot connect to the GMSEC Bus.
		# A GmsecException is thrown if message validation is enabled and a valid Heartbeat message cannot be created.
		#
		# ==== Note
		#
		# If the publish rate is set to 0 (zero), the heartbeat generator will only publish one message.
		#
		# ==== See also
		#
		# MistMessage.set_standard_fields
		#
		def initialize(config, hb_msg_subject, hb_pub_rate, fields = nil)
		end


		##
		# Starts the Heartbeat Generator thread.
		#
		# ==== Returns
		#
		# Returns true if the heartbeat generator has successfully started; false otherwise.
		#
		# ==== Note
		# If the publish rate is set to 0 (zero), only one heartbeat message will be published,
		# however the heartbeat generator thread will continue to run.
		#
		# ==== See also
		#
		# HeartbeatGenerator.stop
		#
		def start
		end


		##
		# Stops the heartbeat generator thread.
		#
		# ==== Returns
		#
		# Returns true if the heartbeat generator has been successfully stopped false otherwise.
		#
		# ==== See also
		#
		# HeartbeatGenerator.start
		#
		def stop
		end


		##
		# Reports whether the heartbeat generator is running.
		#
		# ==== Returns
		#
		# Returns true if the heartbeat generator is running; false otherwise.
		#
		def is_running
		end


		##
		# Changes the configured publish rate of the heartbeat generator.
		#
		# ==== Attributes
		#
		# * +hb_pub_rate+ - The new publish rate/interval (in seconds).
		#
		# ==== Note
		#
		# If the publish rate is set to 0 (zero), only one additional heartbeat message will be published.
		#
		def change_publish_rate(hb_pub_rate)
		end


		##
		# Adds the given Field to the heartbeat message.
		#
		# ==== Attributes
		#
		# * +field+ - The Field to add to the heartbeat message.
		#
		# ==== Returns
		#
		# Returns true if an existing field was overwritten; false otherwise.
		#
		# ==== Raises
		#
		# A GmsecException is thrown if for whatever reason the field cannot be added to the message.
		#
		# ==== Note
		#
		# If the PUB-RATE field is provided, then the publish rate for the heartbeat generator will
		# be updated to publish at the (new) rate.
		#
		# ==== Note: set_field is an overloaded method; other versions include:
		#
		# * Method that accepts a field name and an integer field value
		#
		#     set_field(field_name, integer_field_value)
		#
		# * Method that accepts a field name and a floating point field value
		#
		#     set_field(field_name, double_field_value)
		#
		# * Method that accepts a field name and a string field value
		#
		#     set_field(field_name, string_field_value)
		#
		def set_field(field)
		end
	end
end

File: csharp/gmsec4/dox/HeartbeatGenerator.cs
/*
 * Copyright 2007-2021 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */

/* This file is used for generating Doxygen documentation; it only contains
   stubs of constructors and methods associated with the class.
*/


namespace GMSEC.API {

/// <summary>
/// This class can be used to start a thread that will continuously publish C2CX HB messages.
///
/// The Heartbeat Generator creates a C2CX HB message using MistMessage. It then uses a
/// ConnectionManager to publish this message on a periodic basis.
///
/// <seealso cref="Config"/>
/// <seealso cref="ConnectionManager"/>
/// <seealso cref="MistMessage"/>
/// </summary>
public class HeartbeatGenerator : global::System.IDisposable
{
	/// <summary>
	/// Basic class constructor.
	/// </summary>
	///
	/// <param name="config">The Config object to configure the ConnectionManager that is used to publish messages</param>
	/// <param name="hbMsgSubject">The subject to include with the published heartbeat messages</param>
	/// <param name="hbPubRate">The message publish rate (in seconds)</param>
	///
	/// <exception cref="GmsecException">Thrown if the message subject is null or contains an empty string</exception>
	/// <exception cref="GmsecException">Thrown if the ConnectionManager cannot be created or connect to the GMSEC Bus</exception>
	/// <exception cref="GmsecException">Thrown if message validation is enabled and a valid C2CX HB message cannot be created</exception>
	///
	/// <remarks>
	/// <b>Note</b>
	///     - If the publish rate is set to 0 (zero), then the heartbeat generator will only publish one message.
	/// </remarks>
	///
	/// <seealso cref="SetField"/>
	/// <seealso cref="MistMessage.SetStandardFields"/>
	public HeartbeatGenerator(Config config, string hbMsgSubject, ushort hbPubRate) {
	}


	/// <summary>
	/// Class constructor where a list of message fields can be provided.
	/// </summary>
	///
	/// <param name="config">The Config object to configure the ConnectionManager that is used to publish messages</param>
	/// <param name="hbMsgSubject">The subject to include with the published heartbeat messages</param>
	/// <param name="hbPubRate">The message publish rate (in seconds)</param>
	/// <param name="fields">Fields to add to the heartbeat message</param>
	///
	/// <exception cref="GmsecException">Thrown if the message subject is null or contains an empty string</exception>
	/// <exception cref="GmsecException">Thrown if the ConnectionManager cannot be created or connect to the GMSEC Bus</exception>
	/// <exception cref="GmsecException">Thrown if message validation is enabled and a valid C2CX HB message cannot be created</exception>
	///
	/// <remarks>
	/// <b>Note</b>
	///     - If the publish rate is set to 0 (zero), then the heartbeat generator will only publish one message.
	/// </remarks>
	///
	/// <seealso cref="SetField"/>
	/// <seealso cref="MistMessage.SetStandardFields"/>
	public HeartbeatGenerator(Config config, string hbMsgSubject, ushort hbPubRate, FieldList fields) {
	}


	/// <summary>Starts the heartbeat generator thread.</summary>
	///
	/// <returns>True if the heartbeat generator thread has started; false otherwise</returns>
	///
	/// <remarks>
	/// <b>Note</b>
	///     - If the publish rate is set to 0 (zero), only one heartbeat message will be published, and
	///       the heartbeat generator thread will continue to run.
	/// </remarks>
	///
	/// <seealso cref="Stop"/>
	public bool Start() {
	}


	/// <summary>Stops the heartbeat generator thread.</summary>
	///
	/// <returns>True if the heartbeat generator thread has stopped; false otherwise</returns>
	///
	/// <seealso cref="Start"/>
	public bool Stop() {
	}


	/// <summary>Accessor that can be used to query whether the heartbeat generator thread is running.</summary>
	///
	/// <returns>True if the heartbeat generator thread is running; false otherwise</returns>
	///
	/// <seealso cref="Start"/>
	/// <seealso cref="Stop"/>
	public bool IsRunning() {
	}


	/// <summary>Accessor for changing the publish rate of the heartbeat message generator.</summary>
	///
	/// <param name="pubRate">The new publish rate (in seconds)</param>
	///
	/// <remarks>
	/// <b>Note</b>
	///     - If the publish rate is set to 0 (zero), only one additional heartbeat message will be published.
	/// </remarks>
	public void ChangePublishRate(ushort pubRate) {
	}


	/// <summary>Accessor that will add the given field to the heartbeat message.</summary>
	///
	/// <param name="field">The Field to add to the heartbeat message</param>
	///
	/// <returns>True if an existing field was overwritten; false otherwise</returns>
	///
	/// <exception cref="GmsecException">Thrown if for whatever reason the field cannot be added to the message.</exception>
	///
	/// <remarks>
	/// <b>Note</b>
	///     - If the PUB-RATE field is provided, then the publish rate for the heartbeat generator will be updated
	///       to publish at the (new) rate.
	/// </remarks>
	public bool SetField(Field field) {
	}


	/// <summary>Accessor that will smartly add a field to the heartbeat message.</summary>
	///
	/// <param name="fieldName">The name of the field</param>
	/// <param name="fieldValue">The value of the field</param>
	///
	/// <returns>True if an existing field was overwritten; false otherwise</returns>
	///
	/// <exception cref="GmsecException">Thrown if the field name is null or contains an empty string.</exception>
	/// <exception cref="GmsecException">Thrown if for whatever reason the field cannot be added to the message.</exception>
	///
	/// <remarks>
	/// <b>Notes</b>
	///     - This method relies on MistMessage.SetValue(string, long).
	///     - If the PUB-RATE field is provided, then the publish rate for the heartbeat generator will be updated
	///       to publish at the (new) rate.
	/// </remarks>
	public bool SetField(string fieldName, long fieldValue) {
	}


	/// <summary>Accessor that will smartly add a field to the heartbeat message.</summary>
	///
	/// <param name="fieldName">The name of the field</param>
	/// <param name="fieldValue">The value of the field</param>
	///
	/// <returns>True if an existing field was overwritten; false otherwise</returns>
	///
	/// <exception cref="GmsecException">Thrown if the field name is null or contains an empty string.</exception>
	/// <exception cref="GmsecException">Thrown if for whatever reason the field cannot be added to the message.</exception>
	///
	/// <remarks>
	/// <b>Notes</b>
	///     - This method relies on MistMessage.SetValue(string, double).
	///     - If the PUB-RATE field is provided, then the publish rate for the heartbeat generator will be updated
	///       to publish at the (new) rate.
	/// </remarks>
	public bool SetField(string fieldName, double fieldValue) {
	}


	/// <summary>Accessor that will smartly add a field to the heartbeat message.</summary>
	///
	/// <param name="fieldName">The name of the field</param>
	/// <param name="fieldValue">The value of the field</param>
	///
	/// <returns>True if an existing field was overwritten; false otherwise</returns>
	///
	/// <exception cref="GmsecException">Thrown if the field name is null or contains an empty string.</exception>
	/// <exception cref="GmsecException">Thrown if the field value is null.</exception>
	/// <exception cref="GmsecException">Thrown if for whatever reason the field cannot be added to the message.</exception>
	///
	/// <remarks>
	/// <b>Notes</b>
	///     - This method relies on MistMessage.SetValue(string, string).
	///     - If the PUB-RATE field is provided, then the publish rate for the heartbeat generator will be updated
	///       to publish at the (new) rate.
	/// </remarks>
	public bool SetField(string fieldName, string fieldValue) {
	}
}

}
